pkg/squashfs/squashfs.go

```diff
@@ -300,36 +289,310 @@ func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
 	}
 
-	if mounted, err := maybeKernelSquashMount(squashFile, extractDir); err == nil && mounted {
-		return nil
-	} else if err != kernelSquashMountFailed {
+	fdest, err := filepath.Abs(extractDir)
+	if err != nil {
 		return err
 	}
 
-	cmd, err := squashFuse(squashFile, extractDir)
-	if err == nil {
-		if err := cmd.Process.Release(); err != nil {
-			return errors.Errorf("Failed to release process %s: %v", cmd, err)
+	if policy.initialized {
+		if err, ok := policy.Excuses[initName]; ok {
+			return err
+		}
+		mounted, err := policy.Extractor.Mount(squashFile, fdest)
+		if err == nil {
+			if !mounted {
+				return errors.Errorf("%s.Mount(%s, %s) returned unmounted with err=nil",
+					policy.Extractor.Name(), squashFile, fdest)
+			}
+			return nil
 		}
-		return nil
-	} else if err != squashNotFound {
 		return err
 	}
-	if p := which("unsquashfs"); p != "" {
-		log.Debugf("Extracting %s -> %s with unsquashfs -f -d %s %s", extractDir, squashFile, extractDir, squashFile)
-		cmd := exec.Command("unsquashfs", "-f", "-d", extractDir, squashFile)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		cmd.Stdin = nil
-		return cmd.Run()
+
+	// At this point we are the initialzer
+	if policy.Excuses == nil {
+		policy.Excuses = map[string]error{}
+	}
+
+	if policy.Extractors == nil || len(policy.Extractors) == 0 {
+		policy.Excuses[initName] = errors.Errorf("policy had no extractors")
+		return policy.Excuses[initName]
+	}
+
+	var extractor SquashExtractor
+	var mounted bool
+	allExcuses := []string{}
+	for _, extractor = range policy.Extractors {
+		mounted, err = extractor.Mount(squashFile, fdest)
+		if mounted && err == nil {
+			policy.Extractor = extractor
+			log.Debugf("Selected squashfs extractor %s", extractor.Name())
+			return nil
+		} else if !mounted && err == nil {
+			err = errors.Errorf("%s returned unmounted without error", extractor.Name())
+			policy.Excuses[initName] = err
+			return err
+		}
+		policy.Excuses[extractor.Name()] = err
+	}
+
+	for n, exc := range policy.Excuses {
+		allExcuses = append(allExcuses, fmt.Sprintf("%s: %v", n, exc))
+	}
+
+	// nothing worked. populate Excuses[initName]
+	policy.Excuses[initName] = errors.Errorf("No suitable extractor found:\n  " + strings.Join(allExcuses, "\n  "))
+	return policy.Excuses[initName]
+}
+
+//
+// ExtractSingleSquash - extract the squashFile to extractDir
+// Initialize a extractPolicy struct and then call ExtractSingleSquashPolicy
+// wik()th that.
+func ExtractSingleSquash(squashFile string, extractDir string) error {
+	exPolInfo.once.Do(func() {
+		const envName = "STACKER_SQUASHFS_EXTRACT_POLICY"
+		const defPolicy = "kmount squashfuse unsquashfs"
+		val := os.Getenv(envName)
+		if val == "" {
+			val = defPolicy
+		}
+		exPolInfo.policy, exPolInfo.err = NewExtractPolicy(strings.Fields(val)...)
+		if extractPolicyErr == nil {
+			for k, v := range exPolInfo.policy.Excuses {
+				log.Debugf(" squashfs extractor %s is not available: %v", k, v)
+			}
+		}
+	})
```

Comment:
Above is how how the ExtractPolicy gets initialized.  That code only runs once (mutex.once).
After it runs, exPolInfo has filtered out the Extractors where IsAvailable == false.
So we only call IsAvailable once.

Reply:
(I'm not sure whether these comments are replies or not)

This code only checks the AmHostRoot().  So if kernel mount fails for another reason - such as you're host root without CAP_SYS_ADMIN, or you are root in a container that happens to have the full host uid range mapped (but, not being init user ns, has CAP_SYS_ADMIN but not against the init_user_ns and so still cannot mount -t squashfs), then that kernel mount will be attempted every time.

Like I said, that's *probably* ok.

Reply:
Never mind - I'm wrong about this.